### PR TITLE
chore: percy updates

### PR DESCRIPTION
**Pull request by marktnoonan**
- Closes https://cypress-io.atlassian.net/browse/UNIFY-1214

Once this PR is merged, we can enable the "Wait for base build" feature in Percy, because we will ensure `10.0-release` and  `develop` builds always finalize. This will dramatically improve the usability of our Percy diffs.

### Additional details

This PR modifies our circle.yml to:

- Remove Percy target branch and commits, and remove the `required` jobs for percy-finalize. After this PR merges, "wait for base build" will be turned on in our Percy settings, which means each PR will always wait for the ideal build for the target branch to finalize and compare with that, instead of ignoring in-progress or unfinalized builds which often leads to an apparent comparison with `develop` when making a PR into 10.0-release for example.

- Restore the `wait-on-circle-jobs` polling JS process. This polling lets us ensure the that builds on `10.0-release` and `develop` will always finalize, since now PR builds will wait for this before they begin a diff. Without this, builds only finalize when all required jobs pass, which can lead to multiple problems.

- Add new logic to `wait-on-circle-jobs` based on branch names:

Branches `10.0-release` and `develop` will always allow Percy to finalize for a given run.  Even if some jobs that generate snapshots fail in CI, for a given commit on those, we always finalize so that PRs with that commit as the merge base will have something to diff against. The polling allows us to move forward when things are completed, not just when

Other branches - ie, feature branches, will finalize only if all required jobs are passing. If there are flaky tests, PR authors must "rerun from failed" to get a finalized percy build that includes all expected snapshots, or if the flake was introduced on the branch, fixing the flake would be better.

Documentation of how Percy should be used, and where it fits into the PR review process, has been added to the internal Unified Dev Process Google doc, since Percy is only used internally. For the same reason, I've removed the Percy section from the draft Testing Strategy and Styleguide. That guide has been moved into `/guides` at the suggestion of @flotwig, an that seems like the right place for it.

### PR Tasks

- [na] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Review by lmiller1990 — COMMENTED** [review]
Glad we can make Percy more stable. This PR has 229 Percy diffs; is that intentional?

Also what's the best way to review/test this PR?

**Comment by marktnoonan** [comment]
@lmiller1990 

> This PR has 229 Percy diffs; is that intentional?

No exactly, but it expected that we might have poor diffs here: it's a side effect of removing the forced branch/commit for diffs in this branch, while not yet forcing Percy to wait for ideal base builds, and we can't wait for ideal base builds until we merge this code to 10.0-release to ensure those builds will always complete.

> Also what's the best way to review/test this PR?

Mainly I'd like @brian-mann to get a look at the PR to confirm it matches what we talked about Friday -- the only difference is I added `develop` as a branch that should always finalize. @emilyrohrbough and @tbiethman have worked on/around this stuff a lot too so I've tagged you as well. 

Once we merge into 10.0-release, I'll update Percy settings to wait for ideal base builds, and monitor Percy builds on 10.0-release today to make sure they finalize as expected. I'll make a new PR to test that waiting for base builds works out as well.

File: scripts/wait-on-circle-jobs.js
Comment (tbiethman) [inline scripts/wait-on-circle-jobs.js:141]: I know most of this file was just reinstated, but it doesn't look like this function is used anymore. Can we remove this and anything else that's not actively used?

Comment (marktnoonan) [reply scripts/wait-on-circle-jobs.js]: Good point, updated 👍

**Review by tbiethman — DISMISSED** [review]

File: scripts/wait-on-circle-jobs.js
Comment (emilyrohrbough) [inline scripts/wait-on-circle-jobs.js:108]: Lets log (or something) this so we don't time out when polled jobs take longer than the CircleCi's default 10 minutes. If we don't have output, CircleCi will kill this job.

**Comment by emilyrohrbough** [comment]
@marktnoonan I know this was asked, but 229 new images seem like quite a few. Who would be verifying these are correct? It looks like quite a few were new but I'm not familiar with all of the 10.0.0 changes to know if all of these are correct.

Does this always wait for the latest commit to finish? Or the last successful build on the base branch? We still sometimes experience flake/test failures that could cause this to potentially hang and stall out. 

I'm the one who removed the initial polling job. Oops 🙈 I will say, I'm bummed we need to re-introduce this for Percy to work correctly. Two issues that I remembering being experienced with this setup. 
1. The CircleCI job was often timing after the default 10 minutes due to lack of console output because one or more of the jobs it was polling on a job that was taking longer than 10 minutes to finish.
2. The flow / progression on how to get to a passing CI state was essential tribal knowledge and hard to understand if you aren't familiar with the CI setup and the scripts being ran to finalize the Percy jobs. Removing the polling allowed anyone to see this step required 'x,y,z' jobs to finish for the Percy Finalize job to run. Is there way to enhance the polling script to work with the CircleCI  requires feature to make this flow obvious to new/unfamiliar users and outside contributors? Might be something we could follow up on later too.

**Comment by marktnoonan** [comment]
These are good questions. Overall I don't think this is "set and forget", I think we need to monitor closely after merging and updating settings.

> @marktnoonan I know this was asked, but 229 new images seem like quite a few. Who would be verifying these are correct? It looks like quite a few were new but I'm not familiar with all of the 10.0.0 changes to know if all of these are correct. 

Unfortunately the diff for this PR is not going to be consistently usable until after "wait for base build" is turned on, but we can't turn "wait for base build" on until after 10.0-release has the changes from this PR. This PR should have no visual changes at all (or just the 3-4 that we often see due to changing timers in the UI that haven't been blacked out yet).

> Does this always wait for the latest commit to finish? Or the last successful build on the base branch? We still sometimes experience flake/test failures that could cause this to potentially hang and stall out.

This PR itself doesn't change what Percy waits for, but after merge we will turn on the "wait for base build" setting, which will look at the for the common ancestor commit between the feature branch and target branch (if there is a PR open for the feature branch). Percy will hold on to the build it did for the latest commit on the feature branch, and only run the diff if/when the ideal base build exists. This PR is attempting to make builds on `10.0-release` and `develop` always finalize so that the ideal base build should always exist (with the tradeoff that the ideal base build might sometimes be incomplete due to failures).

There are still some situations that might lead to "waiting for based build" statuses to not resolve, but I do think it's better to know the ideal base build didn't finish than to move the base build to a different commit - the further back we go the less the diff reflects the actual changes happening in the PR. Similarly forcing a diff against "latest commit on `10.0-release`" may be slightly misleading if it works (depends how far behind the PR branch is). But it's also sensitive to the timing of that commit's Percy build, if we try to force that diff too soon after 10.0-release updated, we re-enter Percy's logic.

The problem we solved with the temp fix recently wasn't that Percy sometimes picks the "wrong branch" to compare with. I think it has more to do with backmerging commits from `develop` into `10.0-release`. I'll share some stuff with you offline about what I think is going on.

> I'm the one who removed the initial polling job. Oops 🙈 I will say, I'm bummed we need to re-introduce this for Percy to work correctly. Two issues that I remembering being experienced with this setup.

I don't think this was an oops! I wasn't wild about re-introducing the polling either. We only came back to it when it was clear that we wanted different behavior on `10.0-release` and `develop` compared to other branches, and that it's not configurable via the `circle.yml` the way we want.

> 1. The CircleCI job was often timing after the default 10 minutes due to lack of console output because one or more of the jobs it was polling on a job that was taking longer than 10 minutes to finish.

We should keep an eye on this, that does sound familiar - I hadn't seen the changes removing the polling as intended to address that, it looked more like part of wider cleanup. I should have asked what specifically led to that change when I came across it looking at the file's history.

> 2. The flow / progression on how to get to a passing CI state was essential tribal knowledge and hard to understand if you aren't familiar with the CI setup and the scripts being ran to finalize the Percy jobs. 

Agreed, I've added a lot of this knowledge to our internal PR process docs (at least provisionally).

> Removing the polling allowed anyone to see this step required 'x,y,z' jobs to finish for the Percy Finalize job to run.

It's definitely much clearer to look in the yaml. The main problem moving these jobs into required steps in the yaml flipped us from requiring them to _finish_ to requiring them to _pass_, before we would finalize the Percy build on a given CI run. When we had a lot of flake as things were stabilizing, it meant any run that had a failing required job would leave that run's Percy build unfinalized, and we would only finalize on the last CI run that turned things green after a "rerun from failed" - usually just a tiny fraction of the snapshots would be included in the Percy build that was attached to a PR into `10.0-release`. The fix for that was to base the Percy parallel nonce off the commit sha-1 to allow full builds to accumulate be treating reruns as just more parallel runs. This has drawbacks as well though.

> Is there way to enhance the polling script to work with the CircleCI  requires feature to make this flow obvious to new/unfamiliar users and outside contributors? Might be something we could follow up on later too.

I agree we should follow up - this change should be monitored closely/experimented with after merge. Percy doesn't run on external PRs, so in the short short term, documenting this internally will at least let those who make and review PRs that use Percy understand what to expect and how to resolve if something is wrong. In general, for Percy PRs into long running feature branches are more of a challenge to handle correctly than PRs into the default branch, so at a certain point, a lot of this is going to start to "just work" again after `10.0-release` gets merged. 

I do think we should either proactively do something about that 10 min timeout in this PR, or at least have a plan for how to quickly solve it if we do see it being a problem again. Tomorrow I'll review this with Brian and make sure we talk about these things.

File: scripts/wait-on-circle-jobs.js
Comment (marktnoonan) [reply scripts/wait-on-circle-jobs.js:111]: Added 👍

**Review by brian-mann — APPROVED** [review]

**Pull request merged**
